File: drivers/dart/lib/src/redwire/conn.dart
import 'dart:async';
import 'dart:convert';
import 'dart:io';
import 'dart:typed_data';

import '../conn.dart';
import '../errors.dart';
import '../options.dart';
import 'codec.dart';
import 'frame.dart';

/// Subset of `dart:io` Socket APIs that both Socket and SecureSocket
/// implement. Lets us write a single client over either transport.
typedef _Stream = Socket;

/// RedWire client. One instance owns one TCP / TLS socket.
///
/// Operations are serialised internally — callers may invoke methods
/// concurrently but each request waits in line for the previous one
/// to finish receiving its response. Correlation ids are monotonic.
class RedwireConn implements Conn {
  RedwireConn._(this._socket, this._zstd) {
    _subscription = _socket.listen(
      _onData,
      onError: _onError,
      onDone: _onDone,
      cancelOnError: true,
    );
  }

  final _Stream _socket;
  final ZstdCodec? _zstd;
  late final StreamSubscription<Uint8List> _subscription;

  final BytesBuilder _buffer = BytesBuilder(copy: false);
  final List<Completer<Frame>> _waiters = <Completer<Frame>>[];
  final Lock _writeLock = Lock();

  Object? _error;
  bool _eof = false;
  bool _closed = false;
  int _nextCorr = 1;

  /// Open a RedWire connection: TCP (or TLS), magic preamble, Hello /
  /// HelloAck / AuthResponse / AuthOk.
  ///
  /// `host` and `port` are required. `tls` selects TLS vs plaintext.
  static Future<RedwireConn> connect({
    required String host,
    required int port,
    String? token,
    String clientName = 'reddb-dart/0.1.0',
    bool tls = false,
    TlsOptions? tlsOpts,
    Duration timeout = const Duration(seconds: 30),
    ZstdCodec? zstd,
  }) async {
    final socket = await _openSocket(
      host: host,
      port: port,
      tls: tls,
      tlsOpts: tlsOpts,
      timeout: timeout,
    );
    socket.setOption(SocketOption.tcpNoDelay, true);

    final conn = RedwireConn._(socket, zstd);
    try {
      await conn._handshake(
        token: token,
        clientName: clientName,
        timeout: timeout,
      );
      return conn;
    } catch (_) {
      await conn.close();
      rethrow;
    }
  }

  static Future<_Stream> _openSocket({
    required String host,
    required int port,
    required bool tls,
    TlsOptions? tlsOpts,
    required Duration timeout,
  }) async {
    if (!tls) {
      return Socket.connect(host, port, timeout: timeout);
    }
    final ctx = tlsOpts?.context ?? SecurityContext.defaultContext;
    return SecureSocket.connect(
      host,
      port,
      context: ctx,
      onBadCertificate: tlsOpts?.allowInsecure == true ? (_) => true : null,
      supportedProtocols: tlsOpts?.alpnProtocols ?? const ['redwire/1'],
      timeout: timeout,
    );
  }

  Future<void> _handshake({
    required String? token,
    required String clientName,
    required Duration timeout,
  }) async {
    // Magic preamble.
    _socket.add([MAGIC, SUPPORTED_VERSION]);
    await _socket.flush();

    // Hello.
    final methods = token != null ? ['bearer'] : ['anonymous', 'bearer'];
    final hello = utf8.encode(jsonEncode({
      'versions': [SUPPORTED_VERSION],
      'auth_methods': methods,
      'features': 0,
      'client_name': clientName,
    }));
    await _writeFrame(MessageKind.hello, _next(), Uint8List.fromList(hello));

    final ack = await _readFrame(timeout);
    if (ack.kind == MessageKind.authFail) {
      throw AuthRefused(_jsonReason(ack.payload) ?? 'AuthFail at HelloAck');
    }
    if (ack.kind != MessageKind.helloAck) {
      throw ProtocolError(
        'expected HelloAck, got ${MessageKind.name(ack.kind)}',
      );
    }
    final ackParsed = _jsonOf(ack.payload);
    final chosen = ackParsed is Map ? ackParsed['auth'] : null;
    if (chosen is! String) {
      throw ProtocolError('HelloAck missing `auth` field');
    }

    Uint8List respPayload;
    if (chosen == 'anonymous') {
      respPayload = Uint8List(0);
    } else if (chosen == 'bearer') {
      if (token == null) {
        throw AuthRefused(
          'server demanded bearer but no token was supplied',
        );
      }
      respPayload = Uint8List.fromList(
        utf8.encode(jsonEncode({'token': token})),
      );
    } else {
      throw ProtocolError(
        "server picked unsupported auth method: '$chosen'",
      );
    }
    await _writeFrame(MessageKind.authResponse, _next(), respPayload);

    final fin = await _readFrame(timeout);
    if (fin.kind == MessageKind.authFail) {
      throw AuthRefused(_jsonReason(fin.payload) ?? 'auth refused');
    }
    if (fin.kind != MessageKind.authOk) {
      throw ProtocolError(
        'expected AuthOk, got ${MessageKind.name(fin.kind)}',
      );
    }
  }

  // ---------------------------------------------------------------------
  // Public Conn surface
  // ---------------------------------------------------------------------

  @override
  Future<Uint8List> query(String sql) async {
    final resp = await _request(
      MessageKind.query,
      Uint8List.fromList(utf8.encode(sql)),
    );
    return _expectResultOrError(resp);
  }

  @override
  Future<Uint8List> insert(
    String collection,
    Map<String, dynamic> payload,
  ) async {
    final body = utf8.encode(jsonEncode({
      'collection': collection,
      'payload': payload,
    }));
    final resp = await _request(MessageKind.bulkInsert, Uint8List.fromList(body));
    return _expectBulkOrError(resp);
  }

  @override
  Future<Uint8List> bulkInsert(
    String collection,
    List<Map<String, dynamic>> rows,
  ) async {
    final body = utf8.encode(jsonEncode({
      'collection': collection,
      'payloads': rows,
    }));
    final resp = await _request(MessageKind.bulkInsert, Uint8List.fromList(body));
    return _expectBulkOrError(resp);
  }

  @override
  Future<Uint8List> get(String collection, String id) async {
    final body = utf8.encode(jsonEncode({'collection': collection, 'id': id}));
    final resp = await _request(MessageKind.get, Uint8List.fromList(body));
    return _expectResultOrError(resp);
  }

  @override
  Future<Uint8List> delete(String collection, String id) async {
    final body = utf8.encode(jsonEncode({'collection': collection, 'id': id}));
    final resp = await _request(MessageKind.delete, Uint8List.fromList(body));
    if (resp.kind == MessageKind.deleteOk) return resp.payload;
    if (resp.kind == MessageKind.error) {
      throw EngineError(utf8.decode(resp.payload, allowMalformed: true));
    }
    throw ProtocolError(
      'expected DeleteOk/Error, got ${MessageKind.name(resp.kind)}',
    );
  }

  @override
  Future<void> ping() async {
    final resp = await _request(MessageKind.ping, Uint8List(0));
    if (resp.kind != MessageKind.pong) {
      throw ProtocolError(
        'expected Pong, got ${MessageKind.name(resp.kind)}',
      );
    }
  }

  @override
  Future<void> close() async {
    if (_closed) return;
    _closed = true;
    try {
      await _writeFrame(MessageKind.bye, _next(), Uint8List(0));
    } catch (_) {
      // best-effort
    }
    try {
      await _socket.close();
    } catch (_) {
      // ignore
    }
    await _subscription.cancel();
    final err = _error ?? RedDBError('CONNECTION_CLOSED', 'connection closed');
    while (_waiters.isNotEmpty) {
      _waiters.removeAt(0).completeError(err);
    }
  }

  // ---------------------------------------------------------------------
  // Internal helpers
  // ---------------------------------------------------------------------

  int _next() {
    final c = _nextCorr;
    _nextCorr = (_nextCorr + 1) & 0x7FFFFFFFFFFFFFFF;
    return c;
  }

  Future<Frame> _request(int kind, Uint8List payload) async {
    if (_closed) {
      throw RedDBError('CONNECTION_CLOSED', 'connection already closed');
    }
    return _writeLock.synchronized(() async {
      final corr = _next();
      await _writeFrame(kind, corr, payload);
      return _readFrame(const Duration(seconds: 30));
    });
  }

  Future<void> _writeFrame(int kind, int corr, Uint8List payload) async {
    final frame = Frame(kind: kind, correlationId: corr, payload: payload);
    final bytes = encodeFrame(frame, zstd: _zstd);
    _socket.add(bytes);
    await _socket.flush();
  }

  Future<Frame> _readFrame(Duration timeout) {
    if (_error != null) return Future.error(_error!);
    final c = Completer<Frame>();
    _waiters.add(c);
    _tryDeliver();
    if (_eof && !c.isCompleted) {
      _flushWaitersWithEof();
    }
    return c.future.timeout(timeout, onTimeout: () {
      _waiters.remove(c);
      throw RedDBError('TIMEOUT', 'redwire read timed out after $timeout');
    });
  }

  void _onData(Uint8List chunk) {
    _buffer.add(chunk);
    _tryDeliver();
  }

  void _onError(Object err, [StackTrace? st]) {
    _error = err;
    _flushWaitersWithError(err);
  }

  void _onDone() {
    _eof = true;
    _flushWaitersWithEof();
  }

  void _tryDeliver() {
    while (_waiters.isNotEmpty && _buffer.length >= FRAME_HEADER_SIZE) {
      final all = _buffer.toBytes();
      ({Frame frame, int consumed})? r;
      try {
        r = decodeFrame(all, zstd: _zstd);
      } catch (e) {
        _waiters.removeAt(0).completeError(e);
        // Keep remaining bytes — bad frame is unrecoverable but
        // return the rest to the caller's later reads.
        _buffer.clear();
        _buffer.add(all);
        return;
      }
      if (r == null) {
        // Need more bytes — re-buffer the flattened buf so we don't
        // keep flattening repeatedly.
        _buffer.clear();
        _buffer.add(all);
        return;
      }
      final remaining = Uint8List.sublistView(all, r.consumed);
      _buffer.clear();
      if (remaining.isNotEmpty) {
        _buffer.add(remaining);
      }
      _waiters.removeAt(0).complete(r.frame);
    }
  }

  void _flushWaitersWithEof() {
    if (_waiters.isEmpty) return;
    if (_buffer.length > 0) {
      _tryDeliver();
      return;
    }
    final err = _error ?? RedDBError('CONNECTION_CLOSED', 'connection closed');
    _flushWaitersWithError(err);
  }

  void _flushWaitersWithError(Object err) {
    while (_waiters.isNotEmpty) {
      _waiters.removeAt(0).completeError(err);
    }
  }

  Uint8List _expectResultOrError(Frame resp) {
    if (resp.kind == MessageKind.result) return resp.payload;
    if (resp.kind == MessageKind.error) {
      throw EngineError(utf8.decode(resp.payload, allowMalformed: true));
    }
    throw ProtocolError(
      'expected Result/Error, got ${MessageKind.name(resp.kind)}',
    );
  }

  Uint8List _expectBulkOrError(Frame resp) {
    if (resp.kind == MessageKind.bulkOk) return resp.payload;
    if (resp.kind == MessageKind.error) {
      throw EngineError(utf8.decode(resp.payload, allowMalformed: true));
    }
    throw ProtocolError(
      'expected BulkOk/Error, got ${MessageKind.name(resp.kind)}',
    );
  }

  static Object? _jsonOf(Uint8List bytes) {
    if (bytes.isEmpty) return null;
    try {
      return jsonDecode(utf8.decode(bytes));
    } catch (_) {
      return null;
    }
  }

  static String? _jsonReason(Uint8List bytes) {
    final v = _jsonOf(bytes);
    if (v is Map && v['reason'] is String) return v['reason'] as String;
    return null;
  }
}

/// Tiny mutex — Dart doesn't ship one in the SDK and we only need
/// FIFO serialised access for the single socket.
class Lock {
  Future<void> _last = Future.value();

  Future<T> synchronized<T>(Future<T> Function() body) {
    final completer = Completer<T>();
    final prev = _last;
    _last = completer.future
        .then<void>((_) {})
        .catchError((Object _, StackTrace __) {});
    prev.whenComplete(() async {
      try {
        completer.complete(await body());
      } catch (e, st) {
        completer.completeError(e, st);
      }
    });
    return completer.future;
  }
}

File: drivers/go/redwire/conn.go
package redwire

import (
	"context"
	"crypto/tls"
	"crypto/x509"
	"encoding/binary"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net"
	"sync"
	"sync/atomic"
	"time"
)

// Magic + minor-version bytes that prefix every RedWire connection.
const (
	Magic            byte = 0xFE
	SupportedVersion byte = 0x01
)

// AuthMethod selects the credentials a client offers in Hello.
type AuthMethod int

const (
	// AuthAnonymous — server must allow anonymous (auth disabled).
	AuthAnonymous AuthMethod = iota
	// AuthBearer — token (login session or API key).
	AuthBearer
	// AuthScram — username + password via SCRAM-SHA-256 (3 RTTs).
	AuthScram
	// AuthOAuthJWT — JWT bearer issued by an OAuth provider.
	AuthOAuthJWT
)

// AuthCreds carries the secret material for the chosen method. Only the field
// matching Method is consulted.
type AuthCreds struct {
	Method   AuthMethod
	Token    string // bearer
	Username string // scram
	Password string // scram
	JWT      string // oauth-jwt
}

// TLSConfig configures the optional TLS layer.
type TLSConfig struct {
	// ServerName overrides SNI. Defaults to the dial host.
	ServerName string
	// RootCAs is the trusted CA bundle. nil = use system roots.
	RootCAs *x509.CertPool
	// Certificates carries client certs for mTLS.
	Certificates []tls.Certificate
	// InsecureSkipVerify disables cert verification. Dev only.
	InsecureSkipVerify bool
}

// ConnOptions controls the dial + handshake.
type ConnOptions struct {
	Host        string
	Port        int
	Auth        AuthCreds
	ClientName  string
	TLS         *TLSConfig
	DialTimeout time.Duration
}

// Conn is a single, sequential RedWire connection. Reads and writes are
// serialised through a mutex; correlation IDs are monotonically assigned.
type Conn struct {
	raw       net.Conn
	mu        sync.Mutex
	corr      atomic.Uint64
	closed    atomic.Bool
	sessionID string
	username  string
	role      string
	features  uint32
}

// Dial opens a TCP (or TLS) connection, performs the handshake, and returns a
// ready-to-use Conn. ctx controls dial + handshake; once that returns, deadlines
// are managed per call by the operation methods.
func Dial(ctx context.Context, opts ConnOptions) (*Conn, error) {
	addr := fmt.Sprintf("%s:%d", opts.Host, opts.Port)
	dialer := &net.Dialer{}
	if opts.DialTimeout > 0 {
		dialer.Timeout = opts.DialTimeout
	}

	var raw net.Conn
	var err error
	if opts.TLS != nil {
		serverName := opts.TLS.ServerName
		if serverName == "" {
			serverName = opts.Host
		}
		cfg := &tls.Config{
			ServerName:         serverName,
			RootCAs:            opts.TLS.RootCAs,
			Certificates:       opts.TLS.Certificates,
			InsecureSkipVerify: opts.TLS.InsecureSkipVerify,
			NextProtos:         []string{"redwire/1"},
		}
		tlsDialer := &tls.Dialer{NetDialer: dialer, Config: cfg}
		raw, err = tlsDialer.DialContext(ctx, "tcp", addr)
	} else {
		raw, err = dialer.DialContext(ctx, "tcp", addr)
	}
	if err != nil {
		return nil, fmt.Errorf("redwire: dial %s: %w", addr, err)
	}

	c := &Conn{raw: raw}
	c.corr.Store(0)

	// Apply ctx deadline to the handshake if one is set.
	if dl, ok := ctx.Deadline(); ok {
		_ = raw.SetDeadline(dl)
	}

	if err := c.handshake(opts); err != nil {
		_ = raw.Close()
		return nil, err
	}
	// Clear deadline so subsequent calls own it.
	_ = raw.SetDeadline(time.Time{})
	return c, nil
}

// SessionID returns the session id the server assigned during AuthOk.
func (c *Conn) SessionID() string { return c.sessionID }

// Username returns the authenticated username (server-side session label).
func (c *Conn) Username() string { return c.username }

// Role returns the authenticated role string.
func (c *Conn) Role() string { return c.role }

// nextCorr returns a monotonically-increasing correlation id, starting at 1.
func (c *Conn) nextCorr() uint64 {
	return c.corr.Add(1)
}

// applyDeadline enforces ctx cancellation on a single read/write step. It
// returns a cleanup func the caller defers.
func (c *Conn) applyDeadline(ctx context.Context) func() {
	if dl, ok := ctx.Deadline(); ok {
		_ = c.raw.SetDeadline(dl)
		return func() { _ = c.raw.SetDeadline(time.Time{}) }
	}
	// Honour ctx.Done() without a deadline by hard-deadlining briefly when ctx
	// is already cancelled. For an unbounded ctx we just clear deadlines.
	_ = c.raw.SetDeadline(time.Time{})
	return func() {}
}

func (c *Conn) writeAll(p []byte) error {
	for len(p) > 0 {
		n, err := c.raw.Write(p)
		if err != nil {
			return err
		}
		p = p[n:]
	}
	return nil
}

func (c *Conn) readFull(p []byte) error {
	_, err := io.ReadFull(c.raw, p)
	return err
}

func (c *Conn) writeFrame(f *Frame) error {
	enc, err := EncodeFrame(f)
	if err != nil {
		return err
	}
	return c.writeAll(enc)
}

func (c *Conn) readFrame() (*Frame, error) {
	var header [FrameHeaderSize]byte
	if err := c.readFull(header[:]); err != nil {
		return nil, err
	}
	length := binary.LittleEndian.Uint32(header[0:4])
	if length < FrameHeaderSize {
		return nil, fmt.Errorf("redwire: server sent length %d", length)
	}
	if length > MaxFrameSize {
		return nil, fmt.Errorf("redwire: server sent oversized frame %d", length)
	}
	buf := make([]byte, length)
	copy(buf[:FrameHeaderSize], header[:])
	if length > FrameHeaderSize {
		if err := c.readFull(buf[FrameHeaderSize:]); err != nil {
			return nil, err
		}
	}
	frame, _, err := DecodeFrame(buf)
	return frame, err
}

// handshake drives the RedWire negotiation: magic + Hello/HelloAck + auth.
func (c *Conn) handshake(opts ConnOptions) error {
	// 1. Magic + minor version.
	if err := c.writeAll([]byte{Magic, SupportedVersion}); err != nil {
		return fmt.Errorf("redwire: write magic: %w", err)
	}

	// 2. Hello.
	methods := authMethodsForCreds(opts.Auth.Method)
	clientName := opts.ClientName
	if clientName == "" {
		clientName = "reddb-go/0.1"
	}
	helloPayload, err := json.Marshal(map[string]any{
		"versions":     []int{1},
		"auth_methods": methods,
		"features":     0,
		"client_name":  clientName,
	})
	if err != nil {
		return fmt.Errorf("redwire: encode hello: %w", err)
	}
	if err := c.writeFrame(NewFrame(KindHello, c.nextCorr(), helloPayload)); err != nil {
		return fmt.Errorf("redwire: write hello: %w", err)
	}

	// 3. Read HelloAck (or AuthFail).
	ack, err := c.readFrame()
	if err != nil {
		return fmt.Errorf("redwire: read hello-ack: %w", err)
	}
	switch ack.Kind {
	case KindHelloAck:
		// fall through
	case KindAuthFail:
		return fmt.Errorf("redwire: auth refused at HelloAck: %s", parseReason(ack.Payload))
	default:
		return fmt.Errorf("redwire: expected HelloAck, got 0x%02x", ack.Kind)
	}
	chosen, err := parseChosenAuth(ack.Payload)
	if err != nil {
		return err
	}

	// 4. AuthResponse for the chosen method.
	switch chosen {
	case "anonymous":
		if err := c.writeFrame(NewFrame(KindAuthResponse, c.nextCorr(), nil)); err != nil {
			return fmt.Errorf("redwire: write auth-response: %w", err)
		}
	case "bearer":
		if opts.Auth.Method != AuthBearer {
			return errors.New("redwire: server demanded bearer but no token was supplied")
		}
		body, _ := json.Marshal(map[string]any{"token": opts.Auth.Token})
		if err := c.writeFrame(NewFrame(KindAuthResponse, c.nextCorr(), body)); err != nil {
			return fmt.Errorf("redwire: write auth-response: %w", err)
		}
	case "oauth-jwt":
		if opts.Auth.Method != AuthOAuthJWT {
			return errors.New("redwire: server demanded oauth-jwt but no JWT was supplied")
		}
		body, _ := json.Marshal(map[string]any{"jwt": opts.Auth.JWT})
		if err := c.writeFrame(NewFrame(KindAuthResponse, c.nextCorr(), body)); err != nil {
			return fmt.Errorf("redwire: write auth-response: %w", err)
		}
	case "scram-sha-256":
		if opts.Auth.Method != AuthScram {
			return errors.New("redwire: server demanded scram-sha-256 but no credentials were supplied")
		}
		if err := c.runScram(opts.Auth.Username, opts.Auth.Password); err != nil {
			return err
		}
	default:
		return fmt.Errorf("redwire: server picked unsupported auth method: %s", chosen)
	}

	// 5. Read AuthOk / AuthFail (SCRAM path consumes its own AuthOk inside
	// runScram and short-circuits here when it returns nil with sessionID set).
	if c.sessionID == "" {
		final, err := c.readFrame()
		if err != nil {
			return fmt.Errorf("redwire: read auth-ok: %w", err)
		}
		switch final.Kind {
		case KindAuthOk:
			c.applyAuthOk(final.Payload)
		case KindAuthFail:
			return fmt.Errorf("redwire: auth refused: %s", parseReason(final.Payload))
		default:
			return fmt.Errorf("redwire: expected AuthOk, got 0x%02x", final.Kind)
		}
	}
	return nil
}

func (c *Conn) runScram(username, password string) error {
	// SCRAM exchange:
	//   client → server: AuthResponse with client-first
	//   server → client: AuthRequest with server-first
	//   client → server: AuthResponse with client-final
	//   server → client: AuthOk (carries v=server-signature) | AuthFail
	sess, err := NewScramSession(username, password)
	if err != nil {
		return err
	}
	body, _ := json.Marshal(map[string]any{"client_first": sess.ClientFirstMessage()})
	if err := c.writeFrame(NewFrame(KindAuthResponse, c.nextCorr(), body)); err != nil {
		return fmt.Errorf("redwire: scram: write client-first: %w", err)
	}
	// Engine ships server-first as the raw `r=...,s=...,i=...` SCRAM payload
	// inside an AuthRequest frame. Some implementations wrap it in JSON
	// {"server_first": ...} — accept either.
	req, err := c.readFrame()
	if err != nil {
		return fmt.Errorf("redwire: scram: read server-first: %w", err)
	}
	if req.Kind == KindAuthFail {
		return fmt.Errorf("redwire: scram refused: %s", parseReason(req.Payload))
	}
	if req.Kind != KindAuthRequest {
		return fmt.Errorf("redwire: scram: expected AuthRequest, got 0x%02x", req.Kind)
	}
	serverFirstRaw, err := unwrapScramServerFirst(req.Payload)
	if err != nil {
		return err
	}
	sf, err := ParseServerFirst(serverFirstRaw)
	if err != nil {
		return err
	}
	final, am, err := sess.BuildClientFinal(sf)
	if err != nil {
		return err
	}
	finalBody, _ := json.Marshal(map[string]any{"client_final": final})
	if err := c.writeFrame(NewFrame(KindAuthResponse, c.nextCorr(), finalBody)); err != nil {
		return fmt.Errorf("redwire: scram: write client-final: %w", err)
	}
	ok, err := c.readFrame()
	if err != nil {
		return fmt.Errorf("redwire: scram: read auth-ok: %w", err)
	}
	switch ok.Kind {
	case KindAuthOk:
		// Verify server signature when present so a forged AuthOk can't pass.
		if v := extractServerSignature(ok.Payload); v != nil {
			if !VerifyServerSignature([]byte(password), sf.Salt, sf.Iter, am, v) {
				return errors.New("redwire: scram: server signature did not verify")
			}
		}
		c.applyAuthOk(ok.Payload)
		return nil
	case KindAuthFail:
		return fmt.Errorf("redwire: scram refused: %s", parseReason(ok.Payload))
	}
	return fmt.Errorf("redwire: scram: expected AuthOk, got 0x%02x", ok.Kind)
}

func unwrapScramServerFirst(payload []byte) ([]byte, error) {
	// Try JSON {"server_first": "..."} first.
	var asObj map[string]any
	if err := json.Unmarshal(payload, &asObj); err == nil {
		if s, ok := asObj["server_first"].(string); ok {
			return []byte(s), nil
		}
	}
	// Otherwise treat the body as the raw SCRAM string.
	return payload, nil
}

// extractServerSignature pulls the `v` (or `server_signature`) field out of an
// AuthOk JSON payload. Returns nil if not present.
func extractServerSignature(payload []byte) []byte {
	var obj map[string]any
	if err := json.Unmarshal(payload, &obj); err != nil {
		return nil
	}
	for _, key := range []string{"v", "server_signature"} {
		if v, ok := obj[key].(string); ok {
			// Try base64 first (engine ships base64), then hex (spec mentions hex).
			if dec, err := DecodeBase64Std(v); err == nil {
				return dec
			}
			if dec, err := decodeHex(v); err == nil {
				return dec
			}
		}
	}
	return nil
}

func decodeHex(s string) ([]byte, error) {
	if len(s)%2 != 0 {
		return nil, errors.New("hex length odd")
	}
	out := make([]byte, len(s)/2)
	for i := 0; i < len(out); i++ {
		hi, err := hexNibble(s[2*i])
		if err != nil {
			return nil, err
		}
		lo, err := hexNibble(s[2*i+1])
		if err != nil {
			return nil, err
		}
		out[i] = (hi << 4) | lo
	}
	return out, nil
}

func hexNibble(b byte) (byte, error) {
	switch {
	case b >= '0' && b <= '9':
		return b - '0', nil
	case b >= 'a' && b <= 'f':
		return b - 'a' + 10, nil
	case b >= 'A' && b <= 'F':
		return b - 'A' + 10, nil
	}
	return 0, errors.New("bad hex char")
}

func (c *Conn) applyAuthOk(payload []byte) {
	var obj map[string]any
	if err := json.Unmarshal(payload, &obj); err != nil {
		return
	}
	if s, ok := obj["session_id"].(string); ok {
		c.sessionID = s
	} else if s, ok := obj["sub"].(string); ok {
		// Spec uses `sub`; engine emits `session_id`. Accept both.
		c.sessionID = s
	}
	if s, ok := obj["username"].(string); ok {
		c.username = s
	}
	if s, ok := obj["role"].(string); ok {
		c.role = s
	}
	if f, ok := obj["features"].(float64); ok {
		c.features = uint32(f)
	}
}

func authMethodsForCreds(m AuthMethod) []string {
	switch m {
	case AuthBearer:
		return []string{"bearer"}
	case AuthScram:
		return []string{"scram-sha-256"}
	case AuthOAuthJWT:
		return []string{"oauth-jwt"}
	default:
		return []string{"anonymous", "bearer"}
	}
}

func parseChosenAuth(payload []byte) (string, error) {
	var obj map[string]any
	if err := json.Unmarshal(payload, &obj); err != nil {
		return "", fmt.Errorf("redwire: decode hello-ack: %w", err)
	}
	if s, ok := obj["auth"].(string); ok {
		return s, nil
	}
	return "", errors.New("redwire: hello-ack missing 'auth' field")
}

func parseReason(payload []byte) string {
	var obj map[string]any
	if err := json.Unmarshal(payload, &obj); err == nil {
		if s, ok := obj["reason"].(string); ok {
			return s
		}
	}
	if len(payload) > 0 {
		return string(payload)
	}
	return "(no reason)"
}

// --- Operations ------------------------------------------------------

// Query sends a SQL string and returns the raw Result.payload bytes.
func (c *Conn) Query(ctx context.Context, sql string) ([]byte, error) {
	c.mu.Lock()
	defer c.mu.Unlock()
	if c.closed.Load() {
		return nil, errors.New("redwire: connection closed")
	}
	cleanup := c.applyDeadline(ctx)
	defer cleanup()
	if err := c.writeFrame(NewFrame(KindQuery, c.nextCorr(), []byte(sql))); err != nil {
		return nil, err
	}
	resp, err := c.readFrame()
	if err != nil {
		return nil, err
	}
	switch resp.Kind {
	case KindResult:
		return resp.Payload, nil
	case KindError:
		return nil, fmt.Errorf("redwire: query: %s", string(resp.Payload))
	}
	return nil, fmt.Errorf("redwire: query: unexpected kind 0x%02x", resp.Kind)
}

// Insert delivers a single row into the named collection.
func (c *Conn) Insert(ctx context.Context, collection string, payload any) error {
	body, err := json.Marshal(map[string]any{
		"collection": collection,
		"payload":    payload,
	})
	if err != nil {
		return fmt.Errorf("redwire: encode insert: %w", err)
	}
	return c.bulkInsertJSON(ctx, body)
}

// BulkInsert delivers a batch of rows.
func (c *Conn) BulkInsert(ctx context.Context, collection string, rows []any) error {
	body, err := json.Marshal(map[string]any{
		"collection": collection,
		"payloads":   rows,
	})
	if err != nil {
		return fmt.Errorf("redwire: encode bulk_insert: %w", err)
	}
	return c.bulkInsertJSON(ctx, body)
}

func (c *Conn) bulkInsertJSON(ctx context.Context, body []byte) error {
	c.mu.Lock()
	defer c.mu.Unlock()
	if c.closed.Load() {
		return errors.New("redwire: connection closed")
	}
	cleanup := c.applyDeadline(ctx)
	defer cleanup()
	if err := c.writeFrame(NewFrame(KindBulkInsert, c.nextCorr(), body)); err != nil {
		return err
	}
	resp, err := c.readFrame()
	if err != nil {
		return err
	}
	switch resp.Kind {
	case KindBulkOk:
		return nil
	case KindError:
		return fmt.Errorf("redwire: bulk_insert: %s", string(resp.Payload))
	}
	return fmt.Errorf("redwire: bulk_insert: unexpected kind 0x%02x", resp.Kind)
}

// Get fetches one row by id. Returns the raw envelope bytes (`{ok, found, ...}`).
func (c *Conn) Get(ctx context.Context, collection, id string) ([]byte, error) {
	body, _ := json.Marshal(map[string]any{
		"collection": collection,
		"id":         id,
	})
	c.mu.Lock()
	defer c.mu.Unlock()
	if c.closed.Load() {
		return nil, errors.New("redwire: connection closed")
	}
	cleanup := c.applyDeadline(ctx)
	defer cleanup()
	if err := c.writeFrame(NewFrame(KindGet, c.nextCorr(), body)); err != nil {
		return nil, err
	}
	resp, err := c.readFrame()
	if err != nil {
		return nil, err
	}
	switch resp.Kind {
	case KindResult:
		return resp.Payload, nil
	case KindError:
		return nil, fmt.Errorf("redwire: get: %s", string(resp.Payload))
	}
	return nil, fmt.Errorf("redwire: get: unexpected kind 0x%02x", resp.Kind)
}

// Delete removes one row by id.
func (c *Conn) Delete(ctx context.Context, collection, id string) error {
	body, _ := json.Marshal(map[string]any{
		"collection": collection,
		"id":         id,
	})
	c.mu.Lock()
	defer c.mu.Unlock()
	if c.closed.Load() {
		return errors.New("redwire: connection closed")
	}
	cleanup := c.applyDeadline(ctx)
	defer cleanup()
	if err := c.writeFrame(NewFrame(KindDelete, c.nextCorr(), body)); err != nil {
		return err
	}
	resp, err := c.readFrame()
	if err != nil {
		return err
	}
	switch resp.Kind {
	case KindDeleteOk:
		return nil
	case KindError:
		return fmt.Errorf("redwire: delete: %s", string(resp.Payload))
	}
	return fmt.Errorf("redwire: delete: unexpected kind 0x%02x", resp.Kind)
}

// Ping sends a Ping and waits for Pong. Cheap keepalive.
func (c *Conn) Ping(ctx context.Context) error {
	c.mu.Lock()
	defer c.mu.Unlock()
	if c.closed.Load() {
		return errors.New("redwire: connection closed")
	}
	cleanup := c.applyDeadline(ctx)
	defer cleanup()
	if err := c.writeFrame(NewFrame(KindPing, c.nextCorr(), nil)); err != nil {
		return err
	}
	resp, err := c.readFrame()
	if err != nil {
		return err
	}
	if resp.Kind != KindPong {
		return fmt.Errorf("redwire: ping: expected Pong, got 0x%02x", resp.Kind)
	}
	return nil
}

// Close sends Bye (best-effort) and tears down the socket.
func (c *Conn) Close() error {
	if !c.closed.CompareAndSwap(false, true) {
		return nil
	}
	c.mu.Lock()
	defer c.mu.Unlock()
	_ = c.writeFrame(NewFrame(KindBye, c.nextCorr(), nil))
	return c.raw.Close()
}
